src/webgpu/shader/execution/expression/call/builtin/modf.spec.ts

```diff
@@ -32,74 +32,51 @@ import {
 
 export const g = makeTestGroup(GPUTest);
 
-/** @returns an ShaderBuilder that evaluates modf and returns .whole from the result structure */
-function wholeBuilder(): ShaderBuilder {
-  return basicExpressionBuilder(value => `modf(${value}).whole`);
-}
-
-/** @returns an ShaderBuilder that evaluates modf and returns .fract from the result structure */
-function fractBuilder(): ShaderBuilder {
-  return basicExpressionBuilder(value => `modf(${value}).fract`);
-}
-
-/** @returns a fract Case for a given vector input */
-function makeVectorCaseFract(kind: FPKind, v: number[]): Case {
-  const fp = FP[kind];
-  v = v.map(fp.quantize);
-  const fs = v.map(e => {
-    return fp.modfInterval(e).fract;
-  });
-
-  return { input: toVector(v, fp.scalarBuilder), expected: fs };
+function structBuilder(): ShaderBuilder {
+  return basicExpressionBuilder(
+    value => `modf(${value})`,
+    result => `ModfOutput(${result}.fract, ${result}.whole)`
+  );
 }
 
 /** @returns a whole Case for a given vector input */
-function makeVectorCaseWhole(kind: FPKind, v: number[]): Case {
+function makeVectorCase(kind: FPKind, input: number[]): Case {
   const fp = FP[kind];
-  v = v.map(fp.quantize);
-  const ws = v.map(e => {
-    return fp.modfInterval(e).whole;
+  input.map(fp.quantize);
+  const fs: FPInterval[] = [];
+  const ws: FPInterval[] = [];
+  input.forEach(i => {
+    const result = fp.modfInterval(i);
+    fs.push(result.elements[0] as FPInterval);
+    ws.push(result.elements[1] as FPInterval);
   });
 
-  return { input: toVector(v, fp.scalarBuilder), expected: ws };
+  return {
+    input: toVector(input, fp.scalarBuilder),
+    expected: new FPStruct(fp.toVector(fs), fp.toVector(ws)),
+  };
 }
 
 export const d = makeCaseCache('modf', {
-  f32_fract: () => {
+  f32_scalar: () => {
     const makeCase = (n: number): Case => {
       n = FP.f32.quantize(n);
-      return { input: FP.f32.scalarBuilder(n), expected: FP.f32.modfInterval(n).fract };
+      return { input: FP.f32.scalarBuilder(n), expected: FP.f32.modfInterval(n) };
     };
     return fullF32Range().map(makeCase);
   },
-  f32_whole: () => {
-    const makeCase = (n: number): Case => {
-      n = FP.f32.quantize(n);
-      return { input: FP.f32.scalarBuilder(n), expected: FP.f32.modfInterval(n).whole };
-    };
-    return fullF32Range().map(makeCase);
-  },
-  f32_vec2_fract: () => {
-    return vectorF32Range(2).map(makeVectorCaseFract.bind(null, 'f32'));
-  },
-  f32_vec2_whole: () => {
-    return vectorF32Range(2).map(makeVectorCaseWhole.bind(null, 'f32'));
+  f32_vec2: () => {
+    return vectorF32Range(2).map(makeVectorCase.bind(null, 'f32'));
```

Comment:
nit: I think `n => makeVectorCase('f32', n)` is easier to read than `makeVectorCase.bind(null, 'f32')`. It's also shorter.